### DISCUSSION TITLE
Are XML blocks generated by `@code_runner` supposed to contain CDATA section?

### DISCUSSION THREAD

**Discussion by Bekaboo** [olimorris/codecompanion.nvim#376] [Q&A]
AFAIK this plugin use XML for structured response. When using the `@editor` tool, the LLM will response with a XML code block which is then parsed by this plugin to modify the contents of the current buffer. The XML code block must always incode the CDATA section, as mentioned in the prompt here:

https://github.com/olimorris/codecompanion.nvim/blob/77e04c5e6c58dea6bd3bbdc4cc45dd099ba6a38a/lua/codecompanion/helpers/tools/editor.lua#L186

 For the `@code_runner` tool, there is no instruction about the CDATA section:

https://github.com/olimorris/codecompanion.nvim/blob/77e04c5e6c58dea6bd3bbdc4cc45dd099ba6a38a/lua/codecompanion/helpers/tools/code_runner.lua#L41

So at first I thought that the CDATA section is unneeded for this tool. However when trying it I get an error that the XML file is invalid:

![image](https://github.com/user-attachments/assets/38cc4761-123b-422b-8f1e-b1ab4eeee17a)

Only after adding the CDATA sections the XML parser can correctly parse the XML:

![image](https://github.com/user-attachments/assets/85166a14-ef5f-4f07-9f71-9d9bcf0696c5)

So I wonder if we can fix this by adding the missing CDATA instruction in the `@code_runner` tool.

---

Adding "Remember to use CDATA in your XML response." works for me with Claude 3.5 Sonnet as backend:

``````diff
diff --git a/lua/codecompanion/helpers/tools/code_runner.lua b/lua/codecompanion/helpers/tools/code_runner.lua
index 4fde457..80e91c8 100644
--- a/lua/codecompanion/helpers/tools/code_runner.lua
+++ b/lua/codecompanion/helpers/tools/code_runner.lua
@@ -52,6 +52,7 @@ Usage: To use this tool, you need to return an XML markdown code block (with bac
 ```
 
 You must:
+- Remember to use CDATA in your XML response.
 - Only use the tool when prompted by the user. For example "can you update the code for me?" or "can you add ..."
 - Be mindful that you may not be required to use the tool in all of your responses
 - Ensure the code you're executing will be able to parsed as valid XML
``````

After adding this instruction:

![image](https://github.com/user-attachments/assets/6c03e125-b824-4c74-8388-b1d3e577b1ba)

I can open a PR to fix it and provide more detailed information if you wish.

**Comment by olimorris** [comment]
Yep please raise a PR for this. The code_runner tool will be getting a fix up in the next few weeks but for now please do add
